### ISSUE TITLE
zlib CVE-2022-37434

### ISSUE THREAD

**Issue by rvandermeulen** [zlib-ng/zlib-ng#1323]
https://github.com/advisories/GHSA-cfmr-vrgj-vqwv

As far as I can tell, zlib-ng is also affected by this.

**Comment by mtl1979** [comment]
zlib-ng already checks that it doesn't exceed `extra_max`.

**Comment by phprus** [comment]
@mtl1979 
I added the test https://github.com/ivd38/zlib_overflow (line https://github.com/ivd38/zlib_overflow/blob/fb36092a2e6dd81ce5e70d8cc8a1d3b24293697b/inf1.c#L283) into zlib-ng ``infcover.c``  and got the asan error:


```
phprus@mbp asan % ./infcover
2.1.0.devel
inflate init: 7304 allocated
prime: 7304 high water mark
force window allocation: 80304 high water mark
force window replacement: 15280 high water mark
force split window update: 15280 high water mark
use fixed blocks: 7304 high water mark
bad window size: 7304 high water mark
inflate built-in memory routines
inflate bad parameters
bad gzip method: 7304 high water mark
bad gzip flags: 7304 high water mark
bad zlib method: 7304 high water mark
set window size from header: 14608 high water mark
bad zlib window size: 7304 high water mark
check adler32: 7304 high water mark
bad header crc: 7304 high water mark
check gzip length: 7304 high water mark
bad zlib header check: 7304 high water mark
need dictionary: 15280 high water mark
compute adler32: 80304 high water mark
miscellaneous, force memory errors: 7648 high water mark
inflateBack bad parameters
inflateBack bad state: 7304 high water mark
inflateBack built-in memory routines
invalid stored block lengths-late: 7304 high water mark
invalid stored block lengths-back: 7304 high water mark
fixed-late: 7304 high water mark
fixed-back: 7304 high water mark
invalid block type-late: 7304 high water mark
invalid block type-back: 7304 high water mark
stored-late: 40160 high water mark
stored-back: 7304 high water mark
too many length or distance symbols-late: 7304 high water mark
too many length or distance symbols-back: 7304 high water mark
invalid code lengths set-late: 7304 high water mark
invalid code lengths set-back: 7304 high water mark
invalid bit length repeat-late: 7304 high water mark
invalid bit length repeat-back: 7304 high water mark
invalid bit length repeat-late: 7304 high water mark
invalid bit length repeat-back: 7304 high water mark
invalid code -- missing end-of-block-late: 7304 high water mark
invalid code -- missing end-of-block-back: 7304 high water mark
invalid literal/lengths set-late: 7304 high water mark
invalid literal/lengths set-back: 7304 high water mark
invalid distances set-late: 7304 high water mark
invalid distances set-back: 7304 high water mark
invalid literal/length code-late: 7304 high water mark
invalid literal/length code-back: 7304 high water mark
invalid distance code-late: 7304 high water mark
invalid distance code-back: 7304 high water mark
invalid distance too far back-late: 7304 high water mark
invalid distance too far back-back: 7304 high water mark
incorrect data check-late: 7304 high water mark
incorrect length check-late: 7304 high water mark
pull 17-late: 7304 high water mark
pull 17-back: 7304 high water mark
long code-late: 7304 high water mark
long code-back: 7304 high water mark
length extra-late: 40160 high water mark
length extra-back: 7304 high water mark
long distance and extra-late: 40160 high water mark
long distance and extra-back: 7304 high water mark
window end-late: 40160 high water mark
window end-back: 7304 high water mark
inflate_fast TYPE return: 7304 high water mark
window wrap: 15280 high water mark
inflate_table not enough errors
fast length extra bits: 7304 high water mark
fast distance extra bits: 7304 high water mark
fast invalid distance code: 7304 high water mark
fast invalid literal/length code: 7304 high water mark
fast 2nd level codes and too far back: 7304 high water mark
very common case: 15280 high water mark
contiguous and wrap around window: 15280 high water mark
copy direct from output: 7648 high water mark
=================================================================
==22523==ERROR: AddressSanitizer: memcpy-param-overlap: memory ranges [0x000105100bbe,0x000205100bbc) and [0x00010520156a, 0x000205201568) overlap
    #0 0x1032c1ad8 in __asan_memcpy+0xbc (libclang_rt.asan_osx_dynamic.dylib:arm64e+0x3dad8)
    #1 0x102b23408 in zng_inflate inflate.c:555
    #2 0x102b0c770 in inf infcover.c:315
    #3 0x102b0b7d0 in main infcover.c:684
    #4 0x102ec1088 in start+0x204 (dyld:arm64e+0x5088)

0x000105100bbe is located 2 bytes to the right of 12-byte region [0x000105100bb0,0x000105100bbc)
allocated by thread T0 here:
    #0 0x1032c3b58 in wrap_malloc+0x94 (libclang_rt.asan_osx_dynamic.dylib:arm64e+0x3fb58)
    #1 0x102b0c5b4 in inf infcover.c:295
    #2 0x102b0b7d0 in main infcover.c:684
    #3 0x102ec1088 in start+0x204 (dyld:arm64e+0x5088)

0x000105201580 is located 0 bytes to the right of 48-byte region [0x000105201550,0x000105201580)
allocated by thread T0 here:
    #0 0x1032c3ddc in wrap_realloc+0x9c (libclang_rt.asan_osx_dynamic.dylib:arm64e+0x3fddc)
    #1 0x102b0d2bc in h2b infcover.c:267
    #2 0x102b0c67c in inf infcover.c:306
    #3 0x102b0b7d0 in main infcover.c:684
    #4 0x102ec1088 in start+0x204 (dyld:arm64e+0x5088)

SUMMARY: AddressSanitizer: memcpy-param-overlap (libclang_rt.asan_osx_dynamic.dylib:arm64e+0x3dad8) in __asan_memcpy+0xbc
==22523==ABORTING
zsh: abort      ./infcover
```

**Comment by nmoinvaz** [comment]
This is now resolved in develop branch.

**Comment by vldt** [comment]
Is there going to be a new release soon (few weeks) incorporating these changes? Vanilla zlib issued 1.2.13 on Oct 13 with a fix.

If not, is it safe to cherry-pick the fix for this into 2.0.5 or 2.0.6?

**Comment by mtl1979** [comment]
@vldt There is still a few issues to be resolved and pull requests that need to be merged, but hopefully we can have a new release sooner than later. @Dead2 will usually make few days before the release a tracking pull request that lists all the fixed and pending issues on development branch that should be included in the next release.
